sep/sep-021.rst

```diff
@@ -46,62 +58,258 @@ Non-goals:
+  add-on and required settings for other extensions) and enables needed
+  components.
+* ``check_configuration()``: receives the fully-initialized ``Crawler``
+  instance before it starts running, performs additional dependency and
+  configuration requirement checks
 
-Examples::
+Additionally, it may (and should, where appropriate) provide one or more
+variables that can be used for automated detection of possible dependency
+clashes:
 
-    def addon_configure(settings):
-        settings.overrides['DOWNLADER_MIDDLEWARES'].update({
-            'scrapy.contrib.downloadermiddleware.httpcache.HttpCacheMiddleware': 900,
-        })
+* ``REQUIRES``: list of built-in or custom components needed by this add-on, as
+  strings
+* ``MODIFIES``: list of built-in components whose functionality is affected or
+  replaced by this add-on (a custom HTTP cache should list ``httpcache`` here)
+* ``PROVIDES``: list of components provided by this add-on (e.g. ``mongodb``
+  for an extension that provides generic read/write access to a MongoDB
+  database, releasing other components from having to provide their own
+  database access methods)
+
+update_settings()
+-----------------
+
+Called:
+~~~~~~~
+
+Shortly after initialisation of the ``Crawler`` object.
+
+Arguments:
+~~~~~~~~~~
+
+* ``config``: the (parsed) section belong to this add-on read from
+  ``scrapy.cfg``
+* ``settings``: the crawler's ``Settings`` instance containing all project
+  settings
+* ``addon_path``: the path in which Scrapy found this callback; this will allow
+  easily enabling components defined in the same module without having to worry
+  about where it is to be found relative to the Scrapy project
+
+Purpose:
+~~~~~~~~
+
+* Modify ``settings`` to enabled required components
+* Expose some add-on specific configuration (``config``) into the global
+  settings namespace (``settings``) if necessary
+* Raise exception if components can not be properly configured (e.g. on missing
+  dependencies); Scrapy will print this exception *and exit* (making the user
+  explicitly acknowledge that the add-on does not work by forcing her to
+  disable it).
+
+Side note:
+~~~~~~~~~~
+
+The ``MiddlewareManager.from_settings()`` method will receive a slight
+modification to allow directly placing python objects instead of class paths
+in the middleware dict settings. This way, add-ons can place already
+instantiated components into the settings. This allows keeping configuration
+as local to components as possible and avoids cluttering up the global
+settings namespace. Furthermore, it allows reusing components (e.g.  using
+two instances of the same mongodb pipeline to write to different locations).
+
+Examples:
+~~~~~~~~~
 
 ::
 
-    def addon_configure(settings):
+    def update_settings(config, settings, addon_path):
+        # Don't care where this module is located
+        settings.set['DOWNLADER_MIDDLEWARES']({
+            addon_path + '.downloadermw.coolmw': 900,
```

Comment:
This interface was updated, but devs could use `__package__` for example to get the full path for the addon (or just put there a fixed string).